### PR TITLE
drivers: serial: nrfx_uarte: UART shim based on nrfx_uarte driver

### PR DESCRIPTION
**Pull request by nordic-krch**
PR adds new Nordic UART shim which is based on the nrfx_uarte driver. Legacy shim is kept for now and can be used after `CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=y`.

New shim implements only polling and asynchronous API. Interrupt driven API is implemented via generic asynchronous to interrupt driven adaptation layer. New shim is using different approach to RX bytes counting (legacy shim was using RXDRDY interrupt or dedicated TIMER peripheral).

New approach changes the behavior of the RX while being compliant with the API. When there is no RX activity for the specified amount of time UART RX is restarted (stopped with ENDRX_STARTRX short being set). With this approach there is no need for byte counting since DMA transfer ends on timeout and AMOUNT register is updated. However, it means that every `UART_RX_RDY` event comes with the new buffer (offset in `uart_rx_event` is always 0).

PR contains asynchronous to interrupt driven adaptation layer which is using `uart_async_rx` helper added by #63967.

PR is based on #65113 and #63967.

### CODE REVIEW

File: drivers/serial/Kconfig [hunk 1]
```diff
@@ -133,6 +133,20 @@ config UART_ASYNC_RX_HELPER
 	  is delayed. Module implements zero-copy approach with multiple reception
 	  buffers.
 
+config UART_ASYNC_TO_INT_DRIVEN_API
+	bool "Asynchronous to Interrupt driven adaptation layer"
```
Comment (dcpleung) [inline drivers/serial/Kconfig:137]: Maybe skip the prompt so only drivers needing this will enable this kconfig?

**Review by nordicjm — COMMENTED** [review]
See comments in https://github.com/zephyrproject-rtos/zephyr/pull/67446

File: drivers/serial/uart_nrfx_uarte2.c [hunk 2]
```diff
@@ -0,0 +1,1058 @@
...
+	case NRFX_ERROR_BUSY:
+		return -EACCES;
+	case NRFX_ERROR_INVALID_PARAM:
+		return -EINVAL;
+	default:
+		return 0;
+	}
+}
+
+static int uarte_nrfx_pm_action(const struct device *dev,
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:846]: This function should be wrapped in `#ifdef CONFIG_PM_DEVICE`.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 3]
```diff
@@ -0,0 +1,1058 @@
...
+	static struct uarte_nrfx_data CONCAT(uarte_, idx, _data) = { \
+		.a2i_data = IS_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN)) ? \
+			&CONCAT(a2i_data_, idx) : NULL, \
+		.async = (IS_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN)) || \
+			  IS_ENABLED(CONCAT(CONFIG_UART_, idx, _ASYNC))) ? \
+				 &CONCAT(uarte, idx, _async) : NULL \
+	}; \
+	static int CONCAT(uarte_, idx, _init)(const struct device *dev) \
+	{ \
+		(void)uarte_nrfx_pm_action;/* To avoid compilation warnings. */ \
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:1018]: Warnings about unused static functions?

File: drivers/serial/uart_nrfx_uarte2.c [hunk 4]
```diff
@@ -0,0 +1,1058 @@
...
+		      PM_DEVICE_DT_GET(UARTE(idx)), \
+		      &CONCAT(uarte_, idx, _data), \
+		      &CONCAT(uarte_, idx, _config), \
+		      PRE_KERNEL_1, \
+		      CONFIG_KERNEL_INIT_PRIORITY_DEVICE, \
+		      &uart_nrfx_uarte_driver_api)
+
+
+
+#define UARTE_MEMORY_SECTION(idx)					       \
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:1048]: It would be better to locate this macro before `UART_NRF_UARTE_DEVICE`.

**Review by anangl — COMMENTED** [review]
Just a few initial comments. I haven't managed to go through all the code yet.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 5]
```diff
@@ -0,0 +1,1058 @@
...
+	if (timeout != SYS_FOREVER_US) {
+		/* Set minimum interval to 3 RTC ticks. 3 is used due to RTC limitation
+		 * which cannot set timeout for next tick. Assuming delay in processing
+		 * 3 instead of 2 is used. Note that lower value would work in a similar
+		 * way but timeouts would always occur later than expected,  most likely
+		 * after ~3 ticks.
+		 */
+		adata->idle_cnt = RX_TIMEOUT_DIV + 1;
+		adata->rx_timeout = K_USEC(timeout / RX_TIMEOUT_DIV);
+		adata->rx_timeout.ticks = MAX(3, adata->rx_timeout.ticks);
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:464]: We got rid of such rounding up to 3 ticks in 52c55177ba71af328c6069ea56c007e6eb5cdf1d. Why is it coming back? See also https://github.com/nrfconnect/sdk-zephyr/pull/1336#discussion_r1360443426. Kernel ticks may not be equal to RTC ticks, and for GRTC the correction is completely inappropriate.

**Review by anangl — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/serial/uart_async_to_irq.h [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Forward declarations. */
+
+/** @brief Data structure used by the adaptation layer.
+ *
+ * This data must be the first element of the UART device data structure.
```
Comment (anangl) [inline include/zephyr/drivers/serial/uart_async_to_irq.h:30]: A pointer to this data must be that first element. Same for the config.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 7]
```diff
@@ -0,0 +1,1058 @@
...
+		return uarte_nrfx_init(dev); \
+	} \
+	PM_DEVICE_DT_DEFINE(UARTE(idx), uarte_nrfx_pm_action); \
+	IF_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN), \
+		(static int CONCAT(uart_rx_start, idx)(void) { \
+			static const struct device *dev = DEVICE_DT_GET(UARTE(idx)); \
+			start_rx(dev); \
+			return 0; \
+		} \
+		SYS_INIT(CONCAT(uart_rx_start, idx), POST_KERNEL, 0);)) \
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:1036]: Why is this done here, not in `uarte_nrfx_init()`?

File: include/zephyr/drivers/serial/uart_async_to_irq.h [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+ */
+struct uart_async_to_irq_data;
+struct uart_async_to_irq_config;
```
Comment (gmarull) [inline include/zephyr/drivers/serial/uart_async_to_irq.h:33]: only one is documented

File: include/zephyr/drivers/serial/uart_async_to_irq.h [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+void z_uart_async_to_irq_irq_err_disable(const struct device *dev);
+
+/** @internal Interrupt driven pending status function */
+int z_uart_async_to_irq_irq_is_pending(const struct device *dev);
+
+/** @internal Interrupt driven interrupt update function */
+int z_uart_async_to_irq_irq_update(const struct device *dev);
+
+/** @internal Set the irq callback function */
+void z_uart_async_to_irq_irq_callback_set(const struct device *dev,
```
Comment (gmarull) [inline include/zephyr/drivers/serial/uart_async_to_irq.h:257]: do not use `@internal`, guard all entries in `@cond INTERNAL_HIDDEN`

File: drivers/serial/Kconfig.nrfx [hunk 10]
```diff
@@ -26,10 +26,17 @@ config UART_NRFX_UARTE
 	def_bool y
 	depends on DT_HAS_NORDIC_NRF_UARTE_ENABLED
 
+config UART_NRFX_UARTE_LEGACY_SHIM
+	depends on UART_NRFX_UARTE
+	# New shim takes more ROM. Until it is fixed use legacy shim in memory
+	# constraint case.
```
Comment (gmarull) [inline drivers/serial/Kconfig.nrfx:32]: is this expected to improve/be fixed?

File: drivers/serial/Kconfig.nrfx_uart_instance [hunk 11]
```diff
@@ -66,6 +73,43 @@ config UART_$(nrfx_uart_num)_NRF_ASYNC_LOW_POWER
...
+config UART_$(nrfx_uart_num)_A2I_RX_SIZE
+	depends on !UART_NRFX_UARTE_LEGACY_SHIM
+	int "Asynchronous to interrupt driven adaptation layer RX buffer size"
+	default 64 if UART_$(nrfx_uart_num)_INTERRUPT_DRIVEN
+	default 0
+	help
+	  Amount of space dedicated for RX. It is divided into chunks with some
+	  amount of that space used for control data.
+
+config UART_$(nrfx_uart_num)_A2I_RX_BUF_COUNT
```
Comment (gmarull) [inline drivers/serial/Kconfig.nrfx_uart_instance:106]: not a fan is instance-based options, these do not scale well and perpetuate the problem nrfx-based driver have: hardcoded relationship with hw instance numbers. Isn't there a better option?

File: modules/hal_nordic/nrfx/Kconfig [hunk 12]
```diff
@@ -573,6 +573,26 @@ config NRFX_UARTE3
 	depends on $(dt_nodelabel_has_compat,uart3,$(DT_COMPAT_NORDIC_NRF_UARTE))
 	select NRFX_UARTE
 
+config NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+	bool "UARTE GPIO configuration support"
+	default y
+
+config NRFX_UARTE_CONFIG_SKIP_PSEL_CONFIG
+	bool "UARTE PSEL configuration support"
+	default y
```
Comment (gmarull) [inline modules/hal_nordic/nrfx/Kconfig:582]: why is this required? doesn;t driver use pinctrl unconditionally?

File: drivers/serial/uart_nrfx_uarte2.c [hunk 13]
```diff
@@ -0,0 +1,1058 @@
...
+#include <nrfx_uarte.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/linker/devicetree_regions.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/serial/uart_async_to_irq.h>
+#include <haly/nrfy_gpio.h>
```
Comment (gmarull) [inline drivers/serial/uart_nrfx_uarte2.c:25]: please sort headers, local, zephyr, hal/3rd party

File: drivers/serial/Kconfig.nrfx [hunk 14]
```diff
@@ -26,10 +26,17 @@ config UART_NRFX_UARTE
...
+	default y if MCUBOOT
+	bool "Legacy UARTE shim"
```
Comment (anangl) [inline drivers/serial/Kconfig.nrfx:34]: It's better to place the option type and prompt on top, for consistency.

File: modules/hal_nordic/nrfx/Kconfig (hunk 12 above)
Comment (nordic-krch) [reply modules/hal_nordic/nrfx/Kconfig]: we want to tell nrfx to not compile the support for handling GPIO and PSEL in the driver. I changed to so that uart shim is actually implying those options (and removed defaults from here).

File: drivers/serial/Kconfig.nrfx (hunk 10 above)
Comment (nordic-krch) [reply drivers/serial/Kconfig.nrfx:35]: Yes, eventually we want to have one shim only. But harder to say exact when.

File: drivers/serial/uart_nrfx_uarte2.c (hunk 7 above)
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Removed. I think that it got back to times then a2i adaptation layer was using some kernel calls and required that init function is called post kernel but it is not longer the case.

File: include/zephyr/drivers/serial/uart_async_to_irq.h (hunk 6 above)
Comment (nordic-krch) [reply include/zephyr/drivers/serial/uart_async_to_irq.h]: Added. Also added static asserts in the shim to check that.

File: drivers/serial/uart_nrfx_uarte2.c (hunk 5 above)
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Removed.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 15]
```diff
@@ -0,0 +1,1058 @@
...
+};
+
+/* If set then pins are managed when going to low power mode. */
+#define UARTE_CFG_FLAG_GPIO_MGMT   BIT(0)
+
+/* If set then ENDTX is PPI'ed to TXSTOP */
+#define UARTE_CFG_FLAG_PPI_ENDTX   BIT(1)
+
+/* If set then receiver is not used. */
+#define UARTE_CFG_FLAG_NO_RX BIT(3)
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: It's no used after update to use `disable-rx` DT flag.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 16]
```diff
@@ -0,0 +1,1058 @@
...
+ * provided reception buffers on time).
+ */
+static inline uint32_t get_keep_fifo_content_flag(const struct device *dev)
+{
+	struct uarte_nrfx_data *data = dev->data;
+	const struct uarte_nrfx_config *config = dev->config;
+
+	if ((IS_ENABLED(CONFIG_UART_USE_RUNTIME_CONFIGURE) &&
+	     (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) ||
+	     (config->config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) {
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Reworked it so that `uart_config` is in data structure only if runtime configuration is enabled. Otherwise, hwfc is checked in nrfx config structure.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 17]
```diff
@@ -0,0 +1,1058 @@
...
+			.interrupt_priority = DT_IRQ(UARTE(idx), priority), \
+			.config = NRF_UARTE_CONFIG(CONCAT(uart, idx, _config), \
+						UARTE_ODD_PARITY_ALLOWED), \
+			.tx_stop_on_end = \
+				IS_ENABLED(CONCAT(CONFIG_UART_, idx, _ENHANCED_POLL_OUT)), \
+			.skip_psel_cfg = true, \
+			.skip_gpio_cfg = true, \
+		}, \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(UARTE(idx)), \
+		.config = { \
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Removed this structure as it was redundant.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 18]
```diff
@@ -0,0 +1,1058 @@
...
+
+/**
+ * @brief Structure for UARTE configuration.
+ */
+struct uarte_nrfx_config {
+	const struct uart_async_to_irq_config *a2i_config;
+	nrfx_uarte_t nrfx_dev;
+	nrfx_uarte_config_t nrfx_config;
+	const struct pinctrl_dev_config *pcfg;
+	struct uart_config config;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Removed `uart_config` structure for driver configuration structure.

File: drivers/serial/Kconfig.nrfx_uart_instance (hunk 11 above)
Comment (gmarull) [reply drivers/serial/Kconfig.nrfx_uart_instance:106]: ping

File: drivers/serial/uart_nrfx_uarte2.c [hunk 19]
```diff
@@ -0,0 +1,1053 @@
...
+
+	do {
+		/* When runtime PM is used we cannot use early return because then
+		 * we have no information when UART is actually done with the
+		 * transmission. It reduces UART performance however, polling in
+		 * general is not power efficient and should be avoided in low
+		 * power applications.
+		 */
+		err = nrfx_uarte_tx(nrfx_dev, &out_char, 1, NRFX_UARTE_TX_EARLY_RETURN);
+		__ASSERT(err != NRFX_ERROR_INVALID_ADDR, "Invalid address of the buffer");
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:587]: Shouldn't there be something like `if (err != NRFX_ERROR_BUSY) break;` here to avoid potential hanging when assertions are disabled or when an unexpected error (like `NRFX_ERROR_FORBIDDEN`) occurs?

File: drivers/serial/uart_nrfx_uarte2.c [hunk 20]
```diff
@@ -0,0 +1,1053 @@
...
+
+	return 0;
+}
+
+static int start_rx(const struct device *dev)
+{
+	const struct uarte_nrfx_config *cfg = dev->config;
+
+	if (cfg->flags & UARTE_CFG_FLAG_NO_RX) {
+		return -ENOTSUP;
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:775]: `start_rx()` is called at initialization of the device, so this will cause the initialization to fail for instances not using the async API but having the `disable-rx` property set.

File: doc/releases/migration-guide-3.6.rst [hunk 21]
```diff
@@ -175,6 +175,9 @@ Device Drivers and Device Tree
                      <&rcc STM32_SRC_PLL1_Q FDCAN_SEL(1)>;
     };
 
+* Runtime configuration of UART driver is now by default disabled for Nordic UART driver. Motivation
+  for that change is that feature is rarely used and disabling reduces memory footprint.
```
Comment (anangl) [inline doc/releases/migration-guide-3.6.rst:179]: I'd rephrase this a bit. For example: "Runtime configuration is now disabled by default for Nordic UART drivers. The motivation for the change is that this feature is rarely used and disabling it significantly reduces the memory footprint."

File: drivers/serial/uart_nrfx_uarte2.c [hunk 22]
```diff
@@ -0,0 +1,1053 @@
...
+	const struct uarte_nrfx_config *config = dev->config;
+	struct uart_event evt;
+
+	if (event->data.rx.length) {
+		if (data->async->err) {
+			evt.type = UART_RX_STOPPED;
+			evt.data.rx_stop.reason = UARTE_ERROR_FROM_MASK(data->async->err);
+			evt.data.rx_stop.data.buf = event->data.rx.p_buffer;
+			evt.data.rx_stop.data.len = event->data.rx.length;
+			if (!(config->flags & UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API)) {
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Actually I had to change those static inline to macros. That is because not all compilers are smart enough to determine from static inline that there is a dead code. There were linker issues when compiling for nrf52_bsim.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 23]
```diff
@@ -0,0 +1,1053 @@
...
+		 * way but timeouts would always occur later than expected,  most likely
+		 * after ~3 ticks.
+		 */
+		adata->idle_cnt = RX_TIMEOUT_DIV + 1;
+		adata->rx_timeout = K_USEC(timeout / RX_TIMEOUT_DIV);
+	} else {
+		adata->rx_timeout = K_NO_WAIT;
+	}
+
+	atomic_or(&data->flags, UARTE_DATA_FLAG_RX_ENABLED);
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: moved down and setting only on success.

File: drivers/serial/uart_nrfx_uarte2.c (hunk 20 above)
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c:756]: Added error code conversion from -ENOTSUP to 0 in the init function so that -ENOTSUP is also accepted.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 24]
```diff
@@ -0,0 +1,1053 @@
...
+			UARTE_MEMORY_SECTION(idx) __aligned(4);					\
+	static nrfx_uarte_rx_cache_t uarte##idx##_rx_cache_scratch;				\
+	static uint8_t a2i_rx_buf##idx[CONFIG_UART_##idx##_A2I_RX_SIZE];			\
+	PINCTRL_DT_DEFINE(UARTE(idx));								\
+	static const struct uart_async_to_irq_config uarte_a2i_config_##idx =			\
+		UART_ASYNC_TO_IRQ_API_CONFIG_INITIALIZER(&a2i_api,				\
+					  async_to_irq_trampoline,				\
+					  UARTE_PROP(idx, current_speed),			\
+					  uarte##idx##_tx_cache,				\
+					  CONFIG_UART_##idx##_TX_CACHE_SIZE - 1,		\
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte2.c]: Yes, nrfx_uarte when it's getting a TX cache buffer, it is using the last byte in the cache buffer for poll_out so it cannot be used by the a2i layer.

File: modules/hal_nordic/nrfx/Kconfig [hunk 25]
```diff
@@ -573,6 +573,24 @@ config NRFX_UARTE3
...
+config NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+	bool "UARTE GPIO configuration support"
+
+config NRFX_UARTE_CONFIG_SKIP_PSEL_CONFIG
+	bool "UARTE PSEL configuration support"
+
+config NRFX_UARTE_CONFIG_TX_LINK
+	bool "UARTE TX transfer linking support"
+
+config NRFX_UARTE_CONFIG_RX_CACHE_ENABLED
```
Comment (anangl) [inline modules/hal_nordic/nrfx/Kconfig:585]: All these options should have `depends on NRFX_UARTE`.

File: modules/hal_nordic/nrfx/Kconfig [hunk 26]
```diff
@@ -573,6 +573,24 @@ config NRFX_UARTE3
...
+	bool "UARTE RX caching support"
+	depends on $(dt_nodelabel_has_compat,ram3x,$(DT_COMPAT_MMIO_SRAM))
+	default y
```
Comment (anangl) [inline modules/hal_nordic/nrfx/Kconfig:588]: Shouldn't this be rather:
```suggestion
	default y if $(dt_nodelabel_has_compat,ram3x,$(DT_COMPAT_MMIO_SRAM))
```
?
Is there a reason to disallow this option when there is no such node?

File: drivers/serial/uart_nrfx_uarte2.c [hunk 27]
```diff
@@ -0,0 +1,1032 @@
...
+	 : (mask) & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : (mask) & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : (mask) & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+/* Determine if the device has interrupt driven API enabled. */
+#define IS_INT_DRIVEN_API(dev) \
+	(IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) && \
+	 (((const struct uarte_nrfx_config *)dev->config)->flags & \
+	 UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API))
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:176]:
```suggestion
	 (((const struct uarte_nrfx_config *)dev->config)->flags & \
	  UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API))
```

File: drivers/serial/uart_nrfx_uarte2.c [hunk 28]
```diff
@@ -0,0 +1,1032 @@
...
+#define UARTE_ERROR_FROM_MASK(mask)					\
+	((mask) & NRF_UARTE_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : (mask) & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : (mask) & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : (mask) & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+/* Determine if the device has interrupt driven API enabled. */
+#define IS_INT_DRIVEN_API(dev) \
+	(IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) && \
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:174]: In theory, `CONFIG_UART_ASYNC_TO_INT_DRIVEN_API` could be enabled for some other UART driver.
```suggestion
	(UARTE_ANY_INTERRUPT_DRIVEN && \
```

File: drivers/serial/uart_nrfx_uarte2.c [hunk 29]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int stop_rx(const struct device *dev)
+{
+	const struct uarte_nrfx_config *cfg = dev->config;
+
+	if (cfg->flags & UARTE_CFG_FLAG_NO_RX) {
+		return -ENOTSUP;
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:847]: This will make `PM_DEVICE_ACTION_SUSPEND` fail. `stop_rx()` should not be called when RX is disabled as `start_rx()` is not called then.

**Review by anangl — COMMENTED** [review]
LGTM. Just a few nitpicking comments, except for the last one.

File: modules/hal_nordic/nrfx/nrfx_config.h [hunk 30]
```diff
@@ -564,6 +564,22 @@
 #define NRFX_UARTE3_ENABLED 1
 #endif
 
+#ifdef CONFIG_NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+#define NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG 1
+#endif
+
```
Comment (anangl) [inline modules/hal_nordic/nrfx/nrfx_config.h:570]: These blank lines are inconsistent with what we have for other drivers.
And actually these translations and the Kconfig options themselves should be introduced in a separate commit.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 31]
```diff
@@ -0,0 +1,1031 @@
...
+			}
+		}
+		if (!IS_ASYNC_API(dev) && !(cfg->flags & UARTE_CFG_FLAG_NO_RX)) {
+			return start_rx(dev);
+		}
+
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		if ((IS_INT_DRIVEN_API(dev) || IS_POLLING_API(dev)) &&
+		     !(cfg->flags & UARTE_CFG_FLAG_NO_RX)) {
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte2.c:879]: For consistency with `PM_DEVICE_ACTION_RESUME`:
```suggestion
		if (!IS_ASYNC_API(dev) && !(cfg->flags & UARTE_CFG_FLAG_NO_RX)) {
```

**Review by anangl — DISMISSED** [review]

File: drivers/serial/uart_async_to_irq.c [hunk 32]
```diff
@@ -0,0 +1,377 @@
...
+static struct uart_async_to_irq_data *get_data(const struct device *dev)
+{
+	struct uart_async_to_irq_data **data = dev->data;
+
+	return *data;
+}
+
+static const struct uart_async_to_irq_config *get_config(const struct device *dev)
+{
+	const struct uart_async_to_irq_config **config = dev->config;
```
Comment (anangl) [inline drivers/serial/uart_async_to_irq.c:37]: Sorry, I missed that one `const` in my previous comment.
```suggestion
	const struct uart_async_to_irq_config *const *config = dev->config;
```

**Review by anangl — APPROVED** [review]

**Review by gmarull — DISMISSED** [review]
IMHO this should be fixed better https://github.com/zephyrproject-rtos/zephyr/pull/65155/files#r1450353386, but let's move on.

File: drivers/serial/uart_nrfx_uarte2.c [hunk 33]
```diff
@@ -0,0 +1,1030 @@
...
+		 */
+		err = nrfx_uarte_tx(nrfx_dev, &out_char, 1, NRFX_UARTE_TX_EARLY_RETURN);
+		__ASSERT(err != NRFX_ERROR_INVALID_ADDR, "Invalid address of the buffer");
+
+		if (err == NRFX_ERROR_BUSY &&
+		    IS_ENABLED(CONFIG_MULTITHREADING) && k_is_preempt_thread()) {
+			k_msleep(1);
+		}
+		Z_SPIN_DELAY(3);
+	} while (err != NRFX_ERROR_BUSY);
```
Comment (aescolar) [inline drivers/serial/uart_nrfx_uarte2.c:570]: Isn't this a bug? shouldn't it be 
`} while (err != NRFX_SUCCESS);` ?
Otherwise the same byte would be enqueued repeatedly if the Tx is enqueued right away, or?

Comment (anangl) [reply drivers/serial/uart_nrfx_uarte2.c:570]: Yes, it should be `while (err == NRFX_ERROR_BUSY)`.

**Review by anangl — CHANGES_REQUESTED** [review]
See https://github.com/zephyrproject-rtos/zephyr/pull/65155#discussion_r1456033866.

**Review by anangl — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
